### ISSUE TITLE
Multi-line / Block / Inline Comments

### ISSUE THREAD

**Issue by lordkekz** [nushell/nushell#10096]
### Related problem

As far as I can tell, Nushell supports line comments which are started with a `#` but has no syntax for multi-line comments.
This has been discussed before in https://github.com/nushell/nushell/discussions/5734 and https://github.com/nushell/nushell/discussions/6470.

### Describe the solution you'd like

I suggest adding support for inline and multi-line comments. I think the syntax used by Powershell is reasonable:
```powershell
# This is a line comment

<#
This is a block comment
which consists of several lines
#>

$a = 10; <# This is an inline comment #> while ($a -ge 0) { echo $a; $a--; }
```

**Comment by amtoine** [comment]
this sounds like a sensible thing to look into :yum:

**Comment by stellarpower** [comment]
I know some languages argue that multi-line comments aren't needed these days, due to better text editor/IDE support.

Personally, I'm a fan of them, it's nice being able to toggle them on and off quickly, and also, it makes it much quicker to insert snippets into my code - then I can copy and paste them into the shell really quickly if I forget how to use a module I've made.

Also, it's not like they have to span several lines. In my own code, I might insert one inbetween arguments of a function to add a hint of something:

```nu
def concatenateTables [leftTable <# FIXME: need a type annotation here, but what to put? #>, rightTable] {}
```
using the syntax suggested above.

Personally, I find languages that only support comments to the end of a line to be limiting, and this stops me annotating my code quite as well as I'd like, although is not the end of the world.

Also, one thing that's a real nuisance coming from C/C++, is that multi-line comments can't be nested. When you're really debugging or working on some experimental code, I find myself commenting things out, then commenting out further, etc. C especially I don't think had single-line comments in the beginning, so multi-line tend to be used by C programmers frequently on just single lines. It then makes it really hard to comment out a whole section of code (maybe a function I was editing that now doesn't compile, and I need to move onto something else). So I actually started using macros, or people `#if 0` whole blocks.

So if this would be possible to parse, I think it's really useful to allow nesting to an arbitrary depth.

**Comment by cablehead** [comment]
Suggestion from [Doru on discord](https://discord.com/channels/601130461678272522/615329862395101194/1250452624180772876):

```
### repeating could work, and it could be extended for matching an equally sized # at the end, allowing for "nested" comments
I feel like it has basically no precedent though, besides python multiline strings, I dunno if another language does something like that
######
Commented out
###
ignored "nesting"
###
Still comments
######
It's similar to raw nushell strings at least
```

CabalCrow also noted that `/* */` could conflict with parsing globs

**Comment by IanManske** [comment]
Future note: any inline comment syntax we choose needs to be easily distinguishable from raw string syntax `r#''#` .

**Comment by fdncred** [comment]
I could live with 
```
<#
blah 
  blah 
   blah
#>
```

**Comment by kubouch** [comment]
In my view, block comments fall into the category of “nice to have”, but their complexity and other downsides outweigh their benefits.

For example, nesting:
```
<#
foo
<# bar #>
#>
```
Or what if someone has `#>` characters in their comment? We might need to support escapes if we want to use block comments as doc comments, unless we disallow using block comments as doc comments. Or comment delimiters appearing outside of comments, e.g., `let msg = "<# spam"`. These should be easy to handle by the Nushell parser (still requiring design work and being significantly more complex than the basic comments), but code editors would have to support it correctly as well, e.g., tree-sitter grammar for correct highlighting, or block (un)commenting with a shortcut. It could make life harder for tooling developers.

Apart from complexity, multi-line comments are not viewable out of context. Right now, if you see a line in a git diff that doesn't start with `#`, you know it's code. That would no longer be true with multi-line comments.

**Comment by fdncred** [comment]
I'm not a real fan of block comments myself. It's easy enough in most editors to select the lines you want commented and hit the keyboard shortcut to comment them all out.

**Comment by MovGP0** [comment]
Rust uses `//` and `/*  */`, so maybe we should use that.

The `#` and `<#  #>` sign has the advantage of beeing more intuitive for user coming from other shells, but `#` might be used for some language feature in a future version of Rust.

**Comment by fdncred** [comment]
I'm not a fan of // and /* */ since we don't support either of those today.

**Comment by lordkekz** [comment]
> but `#` might be used for some language feature in a future version of Rust.

That shouldn't prevent Nushell from using it differently, right? Or am I missing something?

**Comment by eugenesvk** [comment]
> I'm not a real fan of block comments myself. It's easy enough in most editors to select the lines you want commented and hit the keyboard shortcut to comment them all out.

Then you'd have to do the same for any pasted text, and be careful to only select the block you pasted. And then some text could already have comments, so if you're not careful you can get confused between original comments `#` and the extra ones `#` to avoid data loss.

Then there is always this issue of visual noise that no editor can easily help you with

All of the issues a block comment is exactly designed to address

**Comment by ItsHarper** [comment]
> And then some text could already have comments, so if you're not careful you can get confused between original comments `#` and the extra ones `#` to avoid data loss.

When commenting out a whole block, editors should add a comment indicator to every single line, so that when you un-comment out the block later, already-existing comments remain as comments. Does your editor not do that?

---

To me, the most useful thing about block comments is that you can use them to add comments mid-line. That can't be replicated by single-line comments at all.

**Comment by eugenesvk** [comment]
> so that when you un-comment out the block later, already-existing comments remain as comments

How will you know later which block it was if all the lines look the same?

```
# original comment, not part of a block
# block
# block
# # block, what you meant
```

**Comment by ItsHarper** [comment]
> How will you know later which block it was if all the lines look the same?

There have definitely been times where I've accidentally removed the comment syntax from a real comment, but it became pretty uncommon for me once I got into the habit of doing things in a way that minimizes the chance of mistakes. The main thing is that if I'm commenting out a big block, I'm going to surround that block in blank lines. 

There's two ways I'd consider formatting your example, depending on whether the original comment is relevant outside of the context of the commented-out block:

### Comment belongs with block

```nu
# # original comment, not part of a block
# block
# block
```

### Comment is relevant even ignoring the block

```nu
# original comment, not part of a block

# block
# block
```

**Comment by ItsHarper** [comment]
Also, if I'm commenting out a block that includes blank lines, the editor should be commenting those out as well:

```nu
# original comment, not part of a block

# blockPart1
# blockPart1
#
# blockPart2
# blockPart2
```

That way, it's visually clear what is a part of the commented-out block, and what is not.

**Comment by eugenesvk** [comment]
Sure, bad workarounds exist:

> To me, the most useful thing about block comments is that you can use them to add comments mid-line. That can't be replicated by single-line comments at all.

You can add a new comment line above and add comment text there that aligns with what you wanted to inline-comment (and even add extra spaces in the original line for even better match to what inline comments would look like)


But then it's not "easily select and press a shortcut" anymore ...

**Comment by ItsHarper** [comment]
> Sure, bad workarounds exist

Fair enough, I've got nothing against having block comments as an option. There's pros and cons both ways.

> But then it's not "easily select and press a shortcut" anymore ...

Right. Your suggested approach for an inline comment is good enough for when you want to add documentation, but completely unhelpful for something like commenting out part of an `if` statement's expression:

```nu
if condition1 /*and condition2*/ {}
```
